From 178eaeea613e1f4b933f5c30c150bf6f1458f8b4 Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Fri, 26 Apr 2024 15:11:35 +0800
Subject: [PATCH 22/67] PHYTIUM: net/phytmac: Add version display for phytmac
 driver

Add support for Using the ethtool -i interface or modinfo
phytmac/phytmac_platform command to display driver version
information.

Mainline: NA
Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: I6577196db18b5b71252ed1e890bb997693460b2d

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/8f25123cc1d1a2ee58e4a61da2ba52e423a13ecf
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/net/ethernet/phytium/phytmac.h          |  2 +-
 drivers/net/ethernet/phytium/phytmac_ethtool.c  | 16 ++++++++++++++++
 drivers/net/ethernet/phytium/phytmac_main.c     |  1 +
 drivers/net/ethernet/phytium/phytmac_pci.c      |  1 +
 drivers/net/ethernet/phytium/phytmac_platform.c |  1 +
 5 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/phytium/phytmac.h b/drivers/net/ethernet/phytium/phytmac.h
index 363bc65c72d4..7e6cf22d7b4f 100644
--- a/drivers/net/ethernet/phytium/phytmac.h
+++ b/drivers/net/ethernet/phytium/phytmac.h
@@ -13,7 +13,7 @@

 #define PHYTMAC_DRV_NAME		"phytium-mac"
 #define PHYTMAC_DRV_DESC		"PHYTIUM Ethernet Driver"
-
+#define PHYTMAC_DRIVER_VERSION		"1.0.0"
 #define PHYTMAC_DEFAULT_MSG_ENABLE	  \
 		(NETIF_MSG_DRV		| \
 		NETIF_MSG_PROBE	| \
diff --git a/drivers/net/ethernet/phytium/phytmac_ethtool.c b/drivers/net/ethernet/phytium/phytmac_ethtool.c
index 7cddb01802f2..4f632591142f 100644
--- a/drivers/net/ethernet/phytium/phytmac_ethtool.c
+++ b/drivers/net/ethernet/phytium/phytmac_ethtool.c
@@ -2,6 +2,8 @@

 #include <linux/ethtool.h>
 #include <linux/phy.h>
+#include <linux/pci.h>
+#include <linux/platform_device.h>
 #include "phytmac.h"
 #include "phytmac_v1.h"
 #include "phytmac_v2.h"
@@ -502,6 +504,19 @@ static inline void phytmac_set_msglevel(struct net_device *ndev, u32 level)
 	pdata->msg_enable = level;
 }

+static void phytmac_get_drvinfo(struct net_device *ndev, struct ethtool_drvinfo *drvinfo)
+{
+	struct phytmac *pdata = netdev_priv(ndev);
+
+	strscpy(drvinfo->driver, PHYTMAC_DRV_NAME, sizeof(drvinfo->driver));
+	strscpy(drvinfo->version, PHYTMAC_DRIVER_VERSION, sizeof(drvinfo->version));
+
+	if (pdata->platdev)
+		strscpy(drvinfo->bus_info, pdata->platdev->name, sizeof(drvinfo->bus_info));
+	else if (pdata->pcidev)
+		strscpy(drvinfo->bus_info, pci_name(pdata->pcidev), sizeof(drvinfo->bus_info));
+}
+
 static const struct ethtool_ops phytmac_ethtool_ops = {
 	.get_regs_len			= phytmac_get_regs_len,
 	.get_regs			= phytmac_get_regs,
@@ -524,6 +539,7 @@ static const struct ethtool_ops phytmac_ethtool_ops = {
 	.set_channels			= phytmac_set_channels,
 	.get_wol			= phytmac_get_wol,
 	.set_wol			= phytmac_set_wol,
+	.get_drvinfo			= phytmac_get_drvinfo,
 };

 void phytmac_set_ethtool_ops(struct net_device *ndev)
diff --git a/drivers/net/ethernet/phytium/phytmac_main.c b/drivers/net/ethernet/phytium/phytmac_main.c
index 72b5a15870a0..79170097fd83 100644
--- a/drivers/net/ethernet/phytium/phytmac_main.c
+++ b/drivers/net/ethernet/phytium/phytmac_main.c
@@ -2488,4 +2488,5 @@ MODULE_LICENSE("GPL");
 MODULE_DESCRIPTION("Phytium Ethernet driver");
 MODULE_AUTHOR("Wenting Song");
 MODULE_ALIAS("platform:phytmac");
+MODULE_VERSION(PHYTMAC_DRIVER_VERSION);

diff --git a/drivers/net/ethernet/phytium/phytmac_pci.c b/drivers/net/ethernet/phytium/phytmac_pci.c
index af69329fe06d..60bd296d8f0b 100644
--- a/drivers/net/ethernet/phytium/phytmac_pci.c
+++ b/drivers/net/ethernet/phytium/phytmac_pci.c
@@ -316,3 +316,4 @@ module_pci_driver(phytmac_driver);

 MODULE_LICENSE("GPL");
 MODULE_DESCRIPTION("Phytium NIC PCI wrapper");
+MODULE_VERSION(PHYTMAC_DRIVER_VERSION);
diff --git a/drivers/net/ethernet/phytium/phytmac_platform.c b/drivers/net/ethernet/phytium/phytmac_platform.c
index 305ff5866e2f..9390056fdc7a 100644
--- a/drivers/net/ethernet/phytium/phytmac_platform.c
+++ b/drivers/net/ethernet/phytium/phytmac_platform.c
@@ -253,3 +253,4 @@ MODULE_LICENSE("GPL");
 MODULE_DESCRIPTION("Phytium Ethernet driver");
 MODULE_AUTHOR("Wenting Song");
 MODULE_ALIAS("platform:phytmac");
+MODULE_VERSION(PHYTMAC_DRIVER_VERSION);
--
2.50.1
